nabs/plans.py

```diff
@@ -653,3 +659,130 @@ def daq_a3scan(m1, a1, b1, m2, a2, b2, m3, a3, b3, nsteps):
+    """
+    Scan over two variables in steps simultaneously.
+
+    This is a `nabs` version of ``bluesky``'s built-in
+    `bluesky.plans.list_scan` plan.
+    This scan is designed to be used with 3 motors, the x and y motors that
+    move to a designaget target on a sample wafer, and one scan_motor that can
+    be delay time, laser power or some other motor positions.
+
+    Parameters
+    ----------
+    sample : str
+        The name of the sample we're interested in.
+    detectors : list
+        Objects to read into Python in the scan.
+    x_motor : obj
+        Motor object corresponding to the x axes.
+    y_motor : obj
+        Motor object corresponding to the y axes.
+    scan_motor : obj
+        The motor being scanned. It can be e.g., delay time, laser power, some
+        other motor position, etc.
+    ss : list
+        List of all the points (samples) for the scan_motor to go through.
+    n_shots : int
+        Indicates how many shots should be taken, or how many samples should
+        be scanned on the grid.
+    path : str
+        Path where the sample file is located.
+    """
+    detectors = list(detectors) + [scan_motor]
+
+    xx, yy = get_sample_targets(sample, path)
+    cycle_xx = cycle(xx)
+    cycle_yy = cycle(yy)
+
+    # find the index of the next target to be shot (status is False)
+    next_index = next((index for (index, d) in enumerate(xx)
+                      if d['status'] is False), None)
+    if next_index is None:
+        raise IndexError('Could not get a target index that has not been shot,'
+                         ' probably all targets were shot from this sample?')
+
+    if ((next_index - 1) + (n_shots * len(ss))) >= len(xx):
+        raise IndexError('The number of last shot + n_shots * len(ss): '
+                         f'{next_index - 1} + {n_shots} * {len(ss)}= '
+                         f'{next_index - 1 + (n_shots * len(ss))}: '
+                         f' bigger than the available samples: {len(xx)}. '
+                         'Please provide a number in range.')
+
+    @bpp.run_decorator()
+    def inner_scan():
+        try:
+            for i in range(len(ss)):
+                yield from bps.mv(scan_motor, ss[i])
+                for j in range(n_shots):
+                    x = (next(item['pos']
+                         for item in cycle_xx if item["status"] is False))
+                    y = (next(item['pos']
+                         for item in cycle_yy if item["status"] is False))
+                    yield from bpp.stub_wrapper(bp.list_scan(detectors,
+                                                x_motor, [x], y_motor, [y]))
+            update_sample(sample, path, (n_shots * len(ss)))
+        except Exception:
+            current_position = x_motor.position
+            try:
+                last_index = next((index for (index, d) in enumerate(xx)
+                                  if np.isclose(d["pos"], current_position)))
```

Comment:
I had to use `np.isclose` to try to find the current position in the saved sample. I am using this in case we stop a run, and based on the current motor position I try to see which target has been shot last. This works in theory if the motor's readback position is exactly the same as saved in the file, but this is not the case all the time - how likely is that `np.isclose` will still not be able to match a saved position in the file?

Reply:
Also one more concern here is what if I have two targets that have the same position saved - should I also check if it has been shot?

Reply:
`np.isclose` should return true if the values are within an absolute tolerance of 1e-8 and an relative tolerance of 1e-5: https://numpy.org/doc/stable/reference/generated/numpy.isclose.html#numpy.isclose

If the motor has moved at all, you won't be able to match the position. Otherwise it should be okay. You can tune the atol and rtol parameters to work as you'd like if it turns out to be not okay.

I don't understand the question about having two targets with the same position.